### PR TITLE
Feature/18 led support

### LINKED ISSUE #18: LED support

### PR DESCRIPTION
**Pull request by SquaredPotato**
I may add the other effects in later, just so there's an indication on boot. 
The branch and issue can stay so I or someone else can easily pick this up in the future.

Related to #18

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,6 @@
 cmake_minimum_required(VERSION 3.16)
 
-list(APPEND EXTRA_COMPONENT_DIRS "lib/espp/components" "lib/sdk")
+list(APPEND EXTRA_COMPONENT_DIRS "lib/espp/components" "lib/sdk" "lib/esp-idf-lib/components")
```
Comment (abrondijk) [inline CMakeLists.txt]: `"lib/sdk"` should be added as last directory, to allow us to override any components in espp & esp-idf-lib

File: components/ring_lights/config [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    config LED_STRIP_GPIO
+        int "GPIO number for LED data"
+        default 9
+    config LED_STRIP_NUM
+        int "Number of LED's in strip"
+        default 8
+    config LED_RMT_CHANNEL
+        int "What RMT channel to use for the LED's"
+        default 0
+    choice BOARD_TYPE
```
Comment (abrondijk) [inline components/ring_lights/config]: Shouldn't this be called `LED_TYPE`?

File: components/ring_lights/include/ring_lights.hpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+namespace ring_lights {
+
+#if defined(CONFIG_SK6812)
+	#define LED_TYPE led_strip_type_t::LED_STRIP_SK6812
+#elif defined(CONFIG_WS2812)
+	#define LED_TYPE led_strip_type_t::LED_STRIP_WS2812
+#elif defined(CONFIG_APA106)
+	#define LED_TYPE led_strip_type_t::LED_STRIP_APA106
+#elif defined(SM16703)
```
Comment (abrondijk) [inline components/ring_lights/include/ring_lights.hpp]: Shouldn't this be `CONFIG_SM16703`?

File: components/ring_lights/src/effects.cpp [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+#define M_TAU (2 * M_PI)
+
+#define DEGREE_PER_LED (360.0 / NUM_LEDS)
+#define RAD_PER_LED (M_TAU / NUM_LEDS)
+
+// Reset an angle to between 0 and +inf
+#define WRAP_NEGATIVE_DEGREE(a) (a < 0 ? 360 - a : a)
+
+// Mirror a radian on the right half of the unit circle to the left half
+#define REFLECT_ON_Y_RAD(a) (a >= 0 ? M_PI - a : -M_PI - a)
```
Comment (Ricovl) [inline components/ring_lights/src/effects.cpp]: This macro is not used anywhere

File: components/ring_lights/src/effects.cpp [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+		if (IS_BETWEEN_A_B_CLOCKWISE_DEGREES(start, correct_end, current_degree)) {
+			buffer[i] = hsv2rgb_rainbow(msg.primary_color);
+		} else {
+			buffer[i] = {{0}, {0}, {0}};
+		}
+	}
+}
+
+void effects::fill(rgb_t (& buffer)[NUM_LEDS], effect_msg& msg) {
+	for (auto& i : buffer) {
```
Comment (Ricovl) [inline components/ring_lights/src/effects.cpp]: In this case, i doesn't really feel like a fitting name as it's mostly used in case of an index.

**Review by Ricovl — CHANGES_REQUESTED** [review]

File: sdkconfig.esp32-s3 [hunk 6]
```diff
@@ -410,10 +410,10 @@ CONFIG_PARTITION_TABLE_MD5=y
 #
 # Compiler options
 #
-CONFIG_COMPILER_OPTIMIZATION_DEFAULT=y
+# CONFIG_COMPILER_OPTIMIZATION_DEFAULT is not set
 # CONFIG_COMPILER_OPTIMIZATION_SIZE is not set
 # CONFIG_COMPILER_OPTIMIZATION_PERF is not set
-# CONFIG_COMPILER_OPTIMIZATION_NONE is not set
+CONFIG_COMPILER_OPTIMIZATION_NONE=y
```
Comment (abrondijk) [inline sdkconfig.esp32-s3]: You gonna optimize it all yourself?

File: sdkconfig.esp32-s3 [hunk 7]
```diff
@@ -1024,7 +1053,7 @@ CONFIG_FATFS_PER_FILE_CACHE=y
 #
 # CONFIG_FREERTOS_SMP is not set
 # CONFIG_FREERTOS_UNICORE is not set
-CONFIG_FREERTOS_HZ=1000
+CONFIG_FREERTOS_HZ=100
```
Comment (abrondijk) [inline sdkconfig.esp32-s3]: This seems undesirable?

File: src/main.cpp [hunk 8]
```diff
@@ -6,53 +6,77 @@
 #include <inttypes.h>
 #include "esp_log.h"
 
-#include "../sdk/manager/include/manager.hpp"
 #include "../sdk/mock_component/include/mock_component.hpp"
-//#include "../sdk/led_strip/include/led_strip.hpp"
+#include "../sdk/manager/include/manager.hpp"
```
Comment (abrondijk) [inline src/main.cpp]: We shouldn't be using relative paths, and the path is not correct as sdk has been moved to `../lib/sdk`. ?¿?

Comment (SquaredPotato) [reply src/main.cpp]: how is this building

File: sdkconfig.esp32-s3 (hunk 6 above)
Comment (SquaredPotato) [reply sdkconfig.esp32-s3]: Yes.

File: components/ring_lights/src/effects.cpp (hunk 5 above)
Comment (SquaredPotato) [reply components/ring_lights/src/effects.cpp]: changing to `led`

**Review by Ricovl — APPROVED** [review]

**Pull request merged**